fetch/fetch.go

```diff
@@ -0,0 +1,526 @@
+		requestList = append(requestList, hash)
+	}
+	f.activeReqM.RUnlock()
+
+	// send in batches
+	for i := 0; i < len(requestList); i += f.cfg.BatchSize {
+		j := i + f.cfg.BatchSize
+		if j > len(requestList) {
+			j = len(requestList)
+		}
+		f.sendBatch(requestList[i:j])
+	}
+}
+
+func (f *Fetch) sendBatch(hashes []types.Hash32) { // create a network batch and serialise to bytes
+	// build list of batch messages
+	var batch requestBatch
+	f.activeReqM.RLock()
+	for _, hash := range hashes {
+		req, ok := f.activeRequests[hash]
+		if !ok {
+			f.log.Warning("message invalidated before sent in batch %v", hash)
+			continue
+		}
+		r := requestMessage{Hint: req[0].hint, Hash: hash}
+		batch.Requests = append(batch.Requests, r)
+	}
+	f.activeReqM.RUnlock()
+	batch.SetID()
+
+	f.activeBatchM.Lock()
+	f.activeBatches[batch.ID] = batch
+	f.activeBatchM.Unlock()
+	// timeout function will be called if no response was received for the hashes sent
+	timeoutFunc := func(err error) {
+		f.handleHashError(batch.ID, err)
+	}
+
+	bytes, err := types.InterfaceToBytes(batch)
+	if err != nil {
+		f.handleHashError(batch.ID, err)
+	}
+	// try sending batch to some random peer
+	retries := 0
+	for {
+		// get random peer
+		p := f.net.GetRandomPeer()
+		err := f.net.SendRequest(fetch, bytes, p, f.receiveResponse, timeoutFunc)
+		// if call succeeded, continue to other Requests
+		if err == nil {
+			break
+		}
+		//todo: mark number of fails per peer to make it low priority
+		retries++
+		if retries > f.cfg.MaxRetiresForPeer {
+			f.handleHashError(batch.ID, ErrCouldNotSend(fmt.Errorf("could not send message")))
+			break
+		}
+		f.log.Error("could not send message to peer %v , retrying, retries : %v", p, retries)
+	}
+}
+
+// handleHashError is called when an error occurred processing batches of the following hashes
+func (f *Fetch) handleHashError(batchHash types.Hash32, err error) {
+	f.log.Error("cannot fetch message %v", err)
+	f.activeBatchM.RLock()
+	batch, ok := f.activeBatches[batchHash]
+	if !ok {
+		f.activeBatchM.RUnlock()
+		f.log.Error("batch invalidated twice %v", batchHash)
+	}
+	f.activeBatchM.RUnlock()
+	f.activeReqM.Lock()
+	for _, h := range batch.Requests {
+		for _, callback := range f.activeRequests[h.Hash] {
+			callback.returnChan <- HashDataPromiseResult{
+				Err:  err,
+				Hash: h.Hash,
+			}
+		}
+		delete(f.activeRequests, h.Hash)
+	}
+	f.activeReqM.Unlock()
+
+	f.activeBatchM.Lock()
+	delete(f.activeBatches, batchHash)
+	f.activeBatchM.Unlock()
+}
+
+// HashDataPromiseResult is the result strict when requesting data corresponding to the Hash
+type HashDataPromiseResult struct {
+	Err  error
+	Hash types.Hash32
+}
+
+// GetAllHashes gets a list of hashes and validates them with the dataCallback function
+// it will return error if any of the hashes were not received.
+func (f *Fetch) GetAllHashes(hashes []types.Hash32, hint Hint, hashValidationFunction HandleDataCallback, validateAndSubmit bool) error {
```

Comment:
GetHashes simplifies the processing of several request with same callback upon result. it performs this callback and return an error if happened. doing it otherwise will make the caller do this logic on his own. also, this method is compatible with the way we currently sync lists. 
Again, if you feel this needs to be decoupled I can do it... it was just my concept on how to handle this data wehen writing it

Reply:
Yes, I think it should be decoupled. Currently the validation is integrated inside of the `Fetcher` internals, while ideally it should be a higher level abstraction which iterates over the results.